guides/dsl.md

```diff
@@ -0,0 +1,289 @@
+
+The interface for wrapping functions for use within Eve is also revamped for the new runtime. When writing a function wrapper, you must ensure that the function is reverentially transparent, meaning given the same input, the function returns the same output.
+
+Functions are wrapped using `makeFunction()` e.g.:
+
+```javascript
+makeFunction({
+  name: "math/sin",             
+  args: {a: "number"},
+  returns: {result: "number"},
+  apply: (a:number) => {
+    return [Math.sin(a/180 * Math.PI)];
+  }
+});
+```
+
+### Aggregates
+
+Aggregates like `sum()`, `count()`, and `sort()` are accessed through the `gather()` function. This function defines the input set to the aggregate. You can optionally group the input set with the `per()` function. For example, here is `count()` at work:
+
+```javascript
+program.bind("count the number of students in each class", ({find, record, gather}) => {
+  let student = find("student")
+  let classSize = gather(student).per(student.teacher).count()
+  return [
+    record("html/element", {tagname: "div", text: `${student.teacher} ${classSize}`})
+  ];
+});
+```
+
+The input set to the `count()` aggregate is the student records, and they are groups according to their teacher attribute. Then, the size of each group is counted and returned in `classSize`, which has the same number of elements as there are teachers.
+
+Aggregates can take arguments as well. For instance, `sort()` takes as arguments the direction by which to sort the input set. Here is `sort()` in use:
+
+```javascript
+program.bind("sort the students by last name, then first name per teacher", ({find, record, gather}) => {
+  let student = find("student")
+  let ix = gather(student.firstName, student.lastName).per(student.teacher).sort("up", "down")
+  return [
+    record("html/element", {tagname: "div", sort: ix, text: `${student.firstName} ${student.lastName}`})
+  ];
+});
+```
+
+## Importing and Exporting Records 
+
+### Getting Data Into Eve - `inputEAVs()`
+
+You can import raw EAVs into Eve with the `inputEAVs()` function. Currently, `inputEAVs()` mut be called at least once to initialize your program, and it can only be used after you define all blocks. These limitations will be lifted in future versions.
+
+`inputEAVs()` takes as input a list of entity, attribute, value triples. The entity value identifies the record to which the attribute and value belong, so it must be unique to that record. For example:
+
+```javascript
+prog.inputEavs([[0,"tag","person"],[0,"name","Archibald"])
+```
+
+This will create a record tagged "person" with the name attribute "Archibald".
+
+### Getting Records out of Eve - Watchers
+
+Watchers are a new feature native to the DSL that allow you to monitor changes in specific records and react to them with a callback function. Like blocks, watchers also attach to a `program`.
```

Comment:
Watcher's aren't specific to the DSL -- They just integrate with whatever method you use for writing Eve blocks. In the future we may support the Eve syntax again or other more exotic ways of creating the watcher's blocks as well. :)